### PR TITLE
Add deferrable mode to dataflow operators

### PR DESCRIPTION
**Pull request by MrGeorgeOwl**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/google/cloud/example_dags/example_dataflow.py
Comment (bhirsz) [inline airflow/providers/google/cloud/example_dags/example_dataflow.py:52]: Those files (in google/cloud/example_dags/) are meant to be migrated to the system tests and then removed. Can you verify if dataflow tests are already migrated and if so modify the migrated file and remove those ones?

Comment (MrGeorgeOwl) [reply airflow/providers/google/cloud/example_dags/example_dataflow.py:52]: System tests was migrated but at the moment I couldn't do required changes even after rebase of the branch

**Comment by potiuk** [comment]
Tests failing.

**Comment by MrGeorgeOwl** [comment]
@potiuk, tests that are failing at the moment are those that comes with changes from main branch. Have the same issues in several PRs so I think that changes in PR are okay but I might be wrong

**Comment by potiuk** [comment]
I think there were intermittent/flaky tests - I re-run them to be sure.

File: airflow/providers/google/cloud/hooks/dataflow.py
Comment (kaxil) [inline airflow/providers/google/cloud/hooks/dataflow.py]: if `environment` accepts `None`, why not add a default of `None`?

```suggestion
    def _update_environment(self, variables: dict,  environment: dict | None = None) -> dict:
```

File: airflow/providers/google/cloud/operators/dataflow.py
Comment (kaxil) [inline airflow/providers/google/cloud/operators/dataflow.py:632]: Can you add this to docstring, please

File: airflow/providers/google/cloud/operators/dataflow.py
Comment (kaxil) [inline airflow/providers/google/cloud/operators/dataflow.py]: If users are inheriting this operator and doing things with `self.hook` it will break their code. Please change it to `self.hook`

File: airflow/providers/google/cloud/operators/dataflow.py
Comment (kaxil) [inline airflow/providers/google/cloud/operators/dataflow.py]:
```suggestion
            self.log.info("Cancelling job %s", self.job_name)
```

File: airflow/providers/google/cloud/operators/dataflow.py
Comment (kaxil) [inline airflow/providers/google/cloud/operators/dataflow.py]: instead you could just make `hook` a  `@cached_property`

File: airflow/providers/google/cloud/operators/dataflow.py
Comment (kaxil) [inline airflow/providers/google/cloud/operators/dataflow.py]: Why make this a required argument?

File: airflow/providers/google/provider.yaml
Comment (kaxil) [inline airflow/providers/google/provider.yaml]: Can you not pin it to a specific version please, and instead provide a range or just min version ideally.

Airflow is a library as well as app so we want to keep deps as open as possible.

File: generated/provider_dependencies.json
Comment (kaxil) [inline generated/provider_dependencies.json]: same as https://github.com/apache/airflow/pull/27776/files#r1054902161

**Comment by brucearctor** [comment]
@MrGeorgeOwl -- this is great and we are eagerly awaiting this functionality to save us some headaches!  Do you want some help getting it over the finish line?  Looks like the outstanding reviews are pretty straightforward :-)

Please advise.

**Comment by MrGeorgeOwl** [comment]
@brucearctor, thank you for your offer of help I appreciate that. I finished required changes at the moment so PR should be merged after it will be reviewed and approved

**Comment by brucearctor** [comment]
Awesome - thanks!

**Comment by MrGeorgeOwl** [comment]
Sorry everyone who was added to this PR as approver. I am not sure why that happened after rebasing to main branch

File: airflow/providers/google/cloud/hooks/dataflow.py
Comment (uranusjr) [inline airflow/providers/google/cloud/hooks/dataflow.py]:
```suggestion
        def _check_one(key, val):
            if key in environment:
                self.log.warning(
                    "%r parameter in 'variables' will override of "
                    "the same one passed in 'environment'!",
                    key,
                )
            return key, val

        environment.update(
            _check_one(key, val)
            for key, val in variables.items()
            if key in environment_keys
        )
```

File: airflow/providers/google/cloud/hooks/dataflow.py
Comment (uranusjr) [inline airflow/providers/google/cloud/hooks/dataflow.py:1057]: Is this related to other changes or just a drive-by improvement?

File: airflow/providers/google/cloud/links/dataflow.py
Comment (uranusjr) [inline airflow/providers/google/cloud/links/dataflow.py:51]: Same question

File: airflow/providers/google/cloud/hooks/dataflow.py
Comment (uranusjr) [inline airflow/providers/google/cloud/hooks/dataflow.py]:
```suggestion
        environment_keys = {
            "numWorkers",
            "maxWorkers",
            "zone",
            "serviceAccountEmail",
            "tempLocation",
            "bypassTempDirValidation",
            "machineType",
            "additionalExperiments",
            "network",
            "subnetwork",
            "additionalUserLabels",
            "kmsKeyName",
            "ipConfiguration",
            "workerRegion",
            "workerZone",
        }
```

Faster for `in` checks.

File: airflow/providers/google/cloud/hooks/dataflow.py
Comment (MrGeorgeOwl) [reply airflow/providers/google/cloud/hooks/dataflow.py:1057]: Drive-by improvement

File: airflow/providers/google/cloud/links/dataflow.py
Comment (MrGeorgeOwl) [reply airflow/providers/google/cloud/links/dataflow.py:51]: Not improvement but fixing the link work, without that change link for Dataflow won't work cause of KeyError exception if I remember correctly

File: airflow/providers/google/cloud/hooks/dataflow.py
Comment (uranusjr) [inline airflow/providers/google/cloud/hooks/dataflow.py]:
```suggestion
                    "%r parameter in 'variables' will override the same one passed in 'environment'!",
```

File: airflow/providers/google/provider.yaml
Comment (uranusjr) [inline airflow/providers/google/provider.yaml:90]: Any reason this version range is so specific? What happens in 0.5.5?

Comment (MrGeorgeOwl) [reply airflow/providers/google/provider.yaml:90]: Version 0.5.5 requires higher versions of protobuf and proto-plus libraries which can break other dependencies in google provider package

Comment (uranusjr) [reply airflow/providers/google/provider.yaml:90]: Typical Google. Thanks for the clarification.

Comment (potiuk) [reply airflow/providers/google/provider.yaml:90]: Can you please add a comment there why ? I know it's not there for the others but we have generally adopted the approach   that whenever we upper-bound a dependency we SHOULD add a comment specifiying why the upper-binding is there and what is the condition for it to be removed. 

https://github.com/apache/airflow#approach-to-dependencies-of-airflow :

> Whenever we upper-bound such a dependency, we should always comment why we are doing it - i.e. we should have a good reason why dependency is upper-bound. And we should also mention what is the condition to remove the binding.

**Review by potiuk — CHANGES_REQUESTED** [review]
Comment why we are adding upper-bound limit

File: airflow/providers/google/provider.yaml
Comment (MrGeorgeOwl) [reply airflow/providers/google/provider.yaml:90]: @potiuk, is comment in provider.yaml will be enough or project has another place for upper-bound comments?

Comment (potiuk) [reply airflow/providers/google/provider.yaml:90]: Yep. Comment in provider.yaml is exactly what I meant

**Comment by MrGeorgeOwl** [comment]
@potiuk, I've added comment about upper-bound for google library

**Review by potiuk — APPROVED** [review]

**Pull request merged**
